src/convex-hull.ts

```diff
@@ -1,71 +1,136 @@
 
 /**
- * 
- * @param outlines 
- * @returns return remaining outlines and the convex hull 
+ * Finds the convex hull of all connected cycles.
+ * When running into an edge case try to set mergeWithTouchingCycles to false.
+ * @param outlines
+ * @param mergeWithTouchingCycles If true, the convex hull will be merged with any of the touching cycles.
+ * @returns return remaining outlines and the convex hull
  */
-export function convexHull(outlines: UnorderdLineSegment[][]): {
-	hull: UnorderdLineSegment[],
-	remainingOutlines: UnorderdLineSegment[][]
+export function convexHull(
+  outlines: UnorderdLineSegment[][],
+  mergeWithTouchingCycles: boolean = true
+): {
+  hull: UnorderdLineSegment[];
+  remainingOutlines: UnorderdLineSegment[][];
 } {
-	const uniquePoints = new UniquePoints();
-	outlines.forEach((outline) => {
-		outline.forEach((line) => {
-			uniquePoints.add(line.start);
-			uniquePoints.add(line.end);
-		});
-	});
-
-	const points = uniquePoints.getPointsSortedByY();
-	const res: Point[] = [];
-	res.push(points[0]);
-	res.push(points[1]);
-
-	let nextPoint;
-
-	while (!nextPoint?.equals(res[0])) {
-		const neighbors = points.filter(
-			(point) => !res[res.length - 2].equals(point) || !res[res.length - 1].equals(point)
-		);
-		if (!neighbors) {
-			break;
-		}
-
-		nextPoint = findPointWithBiggestClockwiseAngle(res[res.length - 2], res[res.length - 1], neighbors);
-
-		res.push(nextPoint);
-	}
-
-	const rawHull = createLinesFromPoints([...res]);
-	const result: UnorderdLineSegment[] = [];
-	const remainingOutlines: UnorderdLineSegment[][] = [];
-
-	outlines.forEach((outline) => {
-		var touchesHull: boolean = false;
-		outline.forEach((line) => {
-			if (includesObject(rawHull, line)) {
-				result.push(...outline);
-				touchesHull = true;
-				return;
-			}
-
-		})
-
-		if (!touchesHull) {
-			remainingOutlines.push(outline);
-		}
-	});
-
-	return {
-		hull: findOutlineOfConnectedCyclesLines([mergeAndRemoveMatchingElements(result, rawHull)]),
-		remainingOutlines: remainingOutlines
-	};
+  // Get all Unique points from the outlines
+  const uniquePoints = new UniquePoints();
+  outlines.forEach((outline) => {
+    outline.forEach((line) => {
+      uniquePoints.add(line.start);
+      uniquePoints.add(line.end);
+    });
+  });
+
+  // Sort the points by y-coordinate to get the first two points
+  const points = uniquePoints.getPointsSortedByY();
+  const pointsOfRawConvexHull: Point[] = [];
+  pointsOfRawConvexHull.push(points[0]);
+  pointsOfRawConvexHull.push(points[1]);
+
+  // Find all points of the initial convex hull
+  let nextPoint;
+  while (!nextPoint?.equals(pointsOfRawConvexHull[0])) {
+    const neighbors = points.filter(
+      (point) =>
+        !pointsOfRawConvexHull[pointsOfRawConvexHull.length - 2].equals(
+          point
+        ) &&
+        !pointsOfRawConvexHull[pointsOfRawConvexHull.length - 1].equals(point)
+    );
+    if (!neighbors) {
```

Comment:
you can not check for empty like this you have to use the length because filter always returns at least an empty array

Reply:
Done